### PR TITLE
CI: Add audit steps for formulae and casks

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Now the steps that say they run on all taps are after the one where we set up all taps. This might have been done on purpose before for performance reasons though. I'm not sure.

### CODE REVIEW

**Comment by apainintheneck** [comment]
```
Error: 338 problems in 102 casks detected
```

Well well well, look what we've got here.

**Comment by apainintheneck** [comment]
Not able to reproduce locally. Probably has something to do with running these audits on Linux and it not being able to decide which `on_*` block to use.

```
~ $ brew audit --eval-all --skip-style --except=version --display-failures-only
==> Auditing a sample of available languages: zh-TW, pt, es-ES, ru, it, es-CL, nl, es-AR, pl and en-
==> Auditing a sample of available languages: cs, nl, fr, zh-TW, ru, ja, it, de, en and en-CA
==> Auditing a sample of available languages: zh, nl, de, en, en-GB, pt, en-CA, gl, zh-TW and it
==> Auditing a sample of available languages: ko, in, gl, es-ES, es-CL, pl, zh-TW, nl, pt and it
==> Auditing a sample of available languages: zh, zh-TW, ja, uk, nl, pt-BR, pl, de, fr and ru
==> Auditing a sample of available languages: bn, gu, eo, fa, he, sr, uk, en-CA, es-ES and zh-TW
==> Auditing a sample of available languages: oc, pa-IN, ar, sr, en-GB, as, nb, nr, mk and af
==> Auditing a sample of available languages: en-GB, pt, uk, ja, nl, cs, pl, en, fr and gl
```

That's probably why it wasn't set up like this before.

**Comment by MikeMcQuaid** [comment]
Thanks @apainintheneck!

Compared to https://github.com/Homebrew/brew/actions/runs/4506520721/jobs/7933410010?pr=15046: this seems to slow things down a lot. I don't think it makes sense to rely on `brew readall` working on Linux for Casks. This could be moved to a (new?) macOS job.

Note that even before this: `tap syntax` was the slowest CI job by a fairly long way so I think it may make sense to split this up into multiple jobs.

Obviously I'd rather we ran more tests more often but, given Linux parallelisation is basically free, I think it makes sense to split out a new job here, perhaps doing just `brew audit`?

**Comment by apainintheneck** [comment]
Looks like running the audit on MacOS pulls in other proprietary taps that have been installed there so I'll need to specify which ones should be included manually. Also, the tap-syntax step still took around 11 minutes even without the audit. If we're running `brew readable` on MacOS it shouldn't be needed on Linux anymore, right?

The new tap-audit step took 16 minutes but that might partially be related to auditing formulae and casks in other taps. I'll rerun it but it might end up being too slow to be worth it. In a pinch we could consider just running the cask stuff on MacOS and the formula stuff on Linux.

**Comment by apainintheneck** [comment]
Still quite slow. It failed after the `homebrew/core` audit and it still took like 10 minutes. Probably not worth it.

**Comment by MikeMcQuaid** [comment]
> In a pinch we could consider just running the cask stuff on MacOS and the formula stuff on Linux.

This would make sense to me!

**Comment by apainintheneck** [comment]
So now it seems to run pretty well after creating two new audit jobs, one for casks on MacOS and another for formulae on Linux.

- [Current Runtime](https://github.com/Homebrew/brew/actions/runs/4520674278/usage)
- [Baseline Runtime](https://github.com/Homebrew/brew/actions/runs/4520010260/usage?pr=15057) - from another recent PR

TLDR: It adds 15 minutes to the total runtime but the two new jobs are run in parallel so it actually takes less time for the user. The tap-syntax job now takes around 10 minutes instead of 12 minutes previously.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -133,6 +122,76 @@ jobs:
...
+        run: brew install-bundler-gems --groups=sorbet
+
+      - name: Set up the homebrew/core tap
+        run: brew tap homebrew/core
+
+      - name: Run brew readall on homebrew/core
+        run: brew readall --aliases homebrew/core
+
+      - name: Run brew audit --skip-style on homebrew/core
+        run: brew audit --skip-style --except=version --display-failures-only --tap=homebrew/core
```
Comment (apainintheneck) [inline .github/workflows/tests.yml:154]: Does it make sense to use `--display-failures-only` here? We used to only use it for the general audit and not the one of `homebrew/core`. I'm going to rerun it without this option and see if it takes longer to complete.

Comment (apainintheneck) [reply .github/workflows/tests.yml]: Seems to run at the same speed so I think it should be fine without it.

Comment (apainintheneck) [reply .github/workflows/tests.yml]: It does end up being kind of noisy on the cask side though.

**Comment by MikeMcQuaid** [comment]
@apainintheneck Great work! We've historically had some issues with macOS worker parallelisation but these seem to be gone now. Just a heads up in case we need to adjust this in future.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (apainintheneck) [reply .github/workflows/tests.yml:154]: That's an interesting point. I'll remove that commit and then this should be good.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @apainintheneck!
